### PR TITLE
ITF: fake peer on demand OS usage example

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

A test example for new on-demand-os functionality in ITF Fake Peers.

### Benefits

A test example and a proof of work.

### Usage Examples or Tests *[optional]*

Please see `fake_peer_example_test`.

### CODE REVIEW

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 1]
```diff
@@ -314,3 +314,80 @@ TEST_F(FakePeerExampleFixture,
...
+ * \attention this code is nothing more but an example of Fake Peer usage
+ *
+ * @given a network of two iroha peers
+ * @when a valid command is sent to one
+ * @then it must either (on demand) provide a proposal containing this command,
+ * or request it from the other peer
+ */
+TEST_F(FakePeerExampleFixture,
+       OnDemandOrderingProposalAfterValidCommandReceived) {
+  static constexpr std::chrono::seconds kProposalWaitingTime(1);
```
Comment (nickaleks) [inline test/integration/acceptance/fake_peer_example_test.cpp]: static is useless here

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 2]
```diff
@@ -314,3 +314,80 @@ TEST_F(FakePeerExampleFixture,
...
+
+  /* A custom behaviour that requests a proposal for the round it got vote for,
+   * and if gets one, checks that the proposal contains the given tx hash.
+   */
+  struct CustomBehaviour : public fake_peer::HonestBehaviour {
+    CustomBehaviour(const interface::types::HashType &tx_hash,
+                    bool &got_proposal_from_main_peer)
```
Comment (nickaleks) [inline test/integration/acceptance/fake_peer_example_test.cpp]: If you modify bool member outside of class, don't you need to ensure it is thread safe?

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 3]
```diff
@@ -314,3 +314,80 @@ TEST_F(FakePeerExampleFixture,
...
+ * or request it from the other peer
+ */
+TEST_F(FakePeerExampleFixture,
+       OnDemandOrderingProposalAfterValidCommandReceived) {
+  static constexpr std::chrono::seconds kProposalWaitingTime(1);
+
+  /* A custom behaviour that requests a proposal for the round it got vote for,
+   * and if gets one, checks that the proposal contains the given tx hash.
+   */
+  struct CustomBehaviour : public fake_peer::HonestBehaviour {
```
Comment (nickaleks) [inline test/integration/acceptance/fake_peer_example_test.cpp]: CustomBehaviour is not a very descriptive name.

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 4]
```diff
@@ -314,3 +314,80 @@ TEST_F(FakePeerExampleFixture,
...
+      const auto proposal_from_main_peer = getFakePeer().sendProposalRequest(
+          message->front().hash.vote_round, kProposalWaitingTime);
+      if (proposal_from_main_peer) {
+        got_proposal_from_main_peer_ |= std::any_of(
+            proposal_from_main_peer->transactions().begin(),
+            proposal_from_main_peer->transactions().end(),
+            [this](const auto &tx) { return tx.reducedHash() == tx_hash_; });
+      }
+      HonestBehaviour::processYacMessage(message);
+      //(this->*&fake_peer::HonestBehaviour::processYacMessage)(message);
```
Comment (nickaleks) [inline test/integration/acceptance/fake_peer_example_test.cpp]: is this code needed?

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 5]
```diff
@@ -314,3 +314,80 @@ TEST_F(FakePeerExampleFixture,
...
+    ASSERT_TRUE(std::any_of(
+        block->transactions().begin(),
+        block->transactions().end(),
+        [&hash](const auto &tx) { return tx.reducedHash() == hash; }))
+        << "The block does not contain the transaction!";
+  });
+  ASSERT_TRUE(block_height > 0) << "Did not get last block height value!";
+
+  EXPECT_TRUE(got_proposal_from_fake_peer || got_proposal_from_main_peer)
+      << "The proposal was neither requested from the fake peer, nor served by "
```
Comment (nickaleks) [inline test/integration/acceptance/fake_peer_example_test.cpp]: is this message necessary? Seems like the expression is self explanatory

File: test/integration/acceptance/fake_peer_example_test.cpp (hunk 3 above)
Comment (MBoldyrev) [reply test/integration/acceptance/fake_peer_example_test.cpp]: It is local to this test function, and described at declaration, so I consider a longer descriptive name unneeded here.

File: test/integration/acceptance/fake_peer_example_test.cpp (hunk 2 above)
Comment (MBoldyrev) [reply test/integration/acceptance/fake_peer_example_test.cpp]: done

**Pull request merged**